core/flash/spi_flash_sfdp.c

```diff
@@ -601,7 +601,13 @@ int spi_flash_sfdp_get_4byte_mode_switch (const struct spi_flash_sfdp_basic_tabl
 
 	params = (struct spi_flash_sfdp_basic_parameter_table_1_5*) table->data;
 	if (table->sfdp->sfdp_header.parameter0.minor_revision >= 5) {
-		if ((params->enter_4b & 0x7f) == 0) {
+
+		// 4 bytes only flash, no switch command available:
+		if ((params->table_1_0.dspi_qspi & SPI_FLASH_SFDP_ADDRESS_BYTES) ==
```

Comment:
Is this situation not already covered by the check for enter_4b flags to be 0?  Do you have an example of a flash device that advertises support for enter_4b through SFDP but does in reality does not?  If so, adding that example flash device to the suite of SFDP tests would be useful.

It also seems like this scenario would get handled naturally in the spi_flash layer when it checks for address mode compatibility before actually executing the mode switch.

Reply:
This is not a real flash, it's an OT device, a standalone chip that is on the data path to the flash itself. 
I don't have DS, I only have a sample of the SFDP table that this device provides.
 
Cerberus sees the OT as a regular flash. This device supports only 4 byte address mode. sending any command to change the address mode to 3 or 4 bytes is ignored. The SFDP protocol allows having such a device, though there is no real commercial flash in fixed mode for 4 bytes, AFAIK.

Since those commands are ignored, the device SFDP table doesn't declare them to be part of the command set.  It just declare that it's a fixed 4 byte device, so no need for ENTER4B\EXIT4B commands.

spi_flash_sfdp.h already has such an option:

/**
 * Supported methods for entering and exiting 4-byte addressing mode.
 */
enum spi_flash_sfdp_4byte_addressing {
	SPI_FLASH_SFDP_4BYTE_MODE_UNSUPPORTED,			/**< 4-byte addressing is not supported. */
	SPI_FLASH_SFDP_4BYTE_MODE_COMMAND,				/**< Use a command to switch the mode. */
	SPI_FLASH_SFDP_4BYTE_MODE_COMMAND_WRITE_ENABLE,	/**< Issue write enable before mode switch. */
	SPI_FLASH_SFDP_4BYTE_MODE_FIXED,				/**< Device is permanently in 4-byte mode. */
};

but the last value is not used anywhere.

if spi_flash_sfdp_get_4byte_mode_switch encounter such a device that doesn't support EXIT\ENTER4B but is fixed 4bytes mode, then it shouldn't return SPI_FLASH_SFDP_4BYTE_MODE_UNSUPPORTED, because it does support 4B.

I will add such a unit test and resubmit the patch.

Thanks, Chris, for the review!